### PR TITLE
Refine overhaul

### PR DESCRIPTION
**Pull request by secretdataz**
* **Addressed Issue(s)**: #1492
* **Server Mode**: Both
* **Description of Pull Request**: 
This pull request converts refine_db.txt to a new format, refine_db.yml.
This pull request also drops the re-roll behavior used in enriched/HD/blessed refiner NPCs for the new fixed chances.

### CODE REVIEW

File: doc/script_commands.txt [hunk 1]
```diff
@@ -2720,14 +2720,17 @@ Examples:
...
 ---------------------------------------
 
-*getequippercentrefinery(<equipment slot>{,<char_id>})
+*getequippercentrefinery(<equipment slot>{,<enriched>,<char_id>})
 
 This function calculates and returns the percent value chance to successfully
 refine the item found in the specified equipment slot of the invoking character
 by +1. There is no actual formula, the success rate for a given weapon level of
-a certain refine level is found in the db/refine_db.txt file. For a list of
+a certain refine level is found in the db/refine_db.yml file. For a list of
```
Comment (Akkarinage) [inline doc/script_commands.txt]: db/(pre-)re/refine_db.yml

**Comment by aleos89** [comment]
I guess the last bit is to add the Shadow Refine values in here.

**Comment by anacondaqq** [comment]
I'm really against changes for changes (doesnot matter about will it useful or not) but this YML txt format looks damn good. Really easy to read and configure. Btw, are tabs and spaces matters in file?  (like in python) or not?

**Comment by aleos89** [comment]
@anacondaqq Glad you like it! About indentation. Tabs are not supported in YAML. Spacing doesn't matter as long as the child nodes node matches properly. You can find a quick summary on indentation for YAML [here](http://www.yaml.org/spec/current.html#id2519916).

**Comment by Yuchinin** [comment]
Hmm YML... this remind me of minecraft.

**Comment by anacondaqq** [comment]
@Yuchinin, because @secretdataz came to us just from there

**Comment by Yuchinin** [comment]
@anacondaqq Wow from Java to C++? Quite good @secretdataz

**Comment by Yuchinin** [comment]
Oh yeah I just notice a wrong spelling in achievement_db.yml, 'Depdendent' isn't it should be 'Dependent' instead?
Or it is intended?

**Comment by secretdataz** [comment]
@Yuchinin thanks for the compliment :)
Yes, I was in the Minecraft modding and server plugins scene, so I get to experience first hand how YAML is so human-friendly that even 9 years old can edit server configs.

**Comment by Yuchinin** [comment]
@secretdataz I should choose to take the Computer Science instead of Information Management when in my University life that have past. Aww I am old...

**Comment by Mikegyver** [comment]
already tested this PR just now.. but i did Merge branch 'master' into refactor/yaml-refine on my test server.

upon compiling, i got this warning appear.

refine overhaul

and upon starting athena, i got this error.

refine overhaul 2

**Comment by secretdataz** [comment]
@Mikegyver you can revert the last commit to make it work.

**Comment by Mikegyver** [comment]
just tested it with reverting last commit..

i would say.. this is quite nice.. trying on refine using enriched elunium..

manage to refine successfully on 4 equipments refining from +4 to + 7..

enriched ore

glad to see now it is following the official rate especially on special refiner (enriched & beyond + 10).

can't wait for this PR to be merge to master branch.

**Comment by Mikegyver** [comment]
i did git pull on this PR & upon compiling i got warning messages appear.

refine 3

**Comment by secretdataz** [comment]
@Mikegyver try updating to my latest commit.

**Comment by Mikegyver** [comment]
@secretdataz seems to be compiling now without having any warning anymore.

File: doc/script_commands.txt [hunk 2]
```diff
@@ -2737,6 +2740,24 @@ random change of a refine succeeding or failing and then going through with it
...
+
+Valid cost types are:
+
+REFINE_COST_NORMAL     - For normal refining
+REFINE_COST_OVER10     - For refining over +10
+REFINE_COST_HD         - For refining with HD ores
+REFINE_COST_ENRICHED   - For refining with enriched ores
+REFINE_COST_OVER10_HD  - For refining over +10 with HD ores
+
+This function will return zeny required for refining if <information> argument is true.
```
Comment (Atemo) [inline doc/script_commands.txt]: Maybe this is s stupid suggestion : for readability you may use another constant name for information

File: npc/merchants/refine.txt [hunk 3]
```diff
@@ -562,27 +562,27 @@ lhz_in02,282,20,7	script	Fulerr	869,{
...
+	.@features = getarg(1);
 	mes "["+ .@npc_name$ +"]";
 	mes "I'm the Armsmith.";
 	mes "I can refine all kinds of weapons, armor and equipment, so let me";
 	mes "know what you want me to refine.";
 	next;
 
 	setarray .@indices[1], EQI_HEAD_TOP, EQI_ARMOR, EQI_HAND_L, EQI_HAND_R, EQI_GARMENT, EQI_SHOES, EQI_ACC_L, EQI_ACC_R, EQI_HEAD_MID, EQI_HEAD_LOW;
-	for(set .@i,1; .@i<getarraysize(.@indices); set .@i,.@i+1) {
+	for(.@i = 1; .@i<getarraysize(.@indices); set .@i,.@i+1) {
```
Comment (Atemo) [inline npc/merchants/refine.txt]: `for(.@i = 1; .@i < getarraysize(.@indices); .@i++) {`

File: db/re/refine_db.yml [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+    - Level: 14
+      NormalChance: 18
+      EnrichedChance: 18
+      EventNormalChance: 35
+      EventEnrichedChance: 35
+    - Level: 15
+      NormalChance: 18
+      EnrichedChance: 18
+      EventNormalChance: 35
+      EventEnrichedChance: 35
```
Comment (Mikegyver) [inline db/re/refine_db.yml]: i think it should be 30 not 35.

File: db/re/refine_db.yml [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+      NormalChance: 7
+      EnrichedChance: 7
+      EventNormalChance: 14
+      EventEnrichedChance: 14
+      Bonus: 1200
+    - Level: 18
+      NormalChance: 7
+      EnrichedChance: 7
+      EventNormalChance: 20
+      EventEnrichedChance: 20
```
Comment (Mikegyver) [inline db/re/refine_db.yml]: i think it should be 10. not 20.

**Comment by Mikegyver** [comment]
just want to double confirm. for WeaponLv3, starting +11 to +16, do success EventEnrichedChance are same with with WeaponLv1 & WeaponLv2 (40, 40, 35, 35, 30, 30) or be like these (35, 35, 30, 30, 25, 25)?

**Comment by Mikegyver** [comment]
seems to be working nice... just in my opinion.. it is time to merge into master branch. (i'm not hurrying, but i think once merge into master, a lot more feedback & review can be obtained)...

**Comment by Mikegyver** [comment]
i did run a test again on this PR branch. got no warning / error on compiling.

however when i test on the refine NPC, get error debug message.

refine 4

and in the game, something wrong happen.. see the following screenshot.

screenragnagempak012

screenragnagempak013

screenragnagempak014

**Comment by Mikegyver** [comment]
just tested. they work fine now...

**Comment by mazvi** [comment]
i was test why i can refine without VIP without item need like oridecon,., just need equip and FREE why?

**Pull request merged**